### PR TITLE
[FSDP2] Separate sharded and unsharded gradient dtypes

### LINKED ISSUE #170648: fsdp2 support grad_dtype
### Problem
The `grad_dtype` property of PyTorch tensors, which allows specifying gradient dtype for leaf tensors, is not preserved when applying `torch.distributed.fsdp.fully_shard`. This breaks precision control in distributed training.

### Alternatives

_No response_

### Additional context

_No response_

cc @chauhang @penguinwu

### PR DESCRIPTION
**Pull request by weifengpy**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.15.0) (oldest at bottom):
* __->__ #194434

Fixes #170648. Based on #172074 by @arkadip-maitra.

> Authored with Codex.
>
> Preserve each parameter's pre-FSDP grad_dtype policy on reduced shards. The unsharded leaf uses an explicit MixedPrecisionPolicy.reduce_dtype, otherwise the parameter's explicit grad_dtype (including None); only parameters without an override use the compute dtype. Defaulting unconditionally to compute dtype discarded explicit gradient precision: BF16 parameters with FP32 grad_dtype accumulated 256 + 1 - 256 as 0 instead of 1. Preserving the gradient policy keeps this accumulation in FP32.
>
> Reduce gradients in groups with the same actual dtype while retaining the existing all-gather grouping. Retain every HSDP all-reduce buffer until completion, and preserve higher-precision partial gradients if an unrestricted policy produces a different dtype later. Reduction casts outputs to each shard's dtype, using a flat cast for uniform policies and grouped foreach copies otherwise. Including unused parameters in reduction requires an explicit reduce_dtype when grad_dtype is explicitly None, so every rank can construct compatible zero gradients.
>
> Capture the policy at fully_shard(), including explicit None. Default policies follow parameter dtype changes, while explicit overrides survive conversion and load_state_dict(assign=True) replacement. Direct grad_dtype edits afterward, including before lazy initialization, are intentionally unsupported and raise at the next FSDP boundary. A code comment and API documentation state this limit.
>
> Keep native gradient owners separate: unreduced accumulation stays on the unsharded parameter's .grad, reduced history stays on the sharded parameter's .grad, and HSDP retains local reduction buffers until all-reduce completes. Combining FP32 reduced history with BF16 pending accumulation would lose information: 256 synchronized, then 1 without synchronization, then -256 synchronized must produce 1. HSDP retains per-parameter views so later microbatches can use different parameter subsets without confusing packed offsets.
>
> Start review with dtype initialization and reduction output casts, then native accumulation ownership and gradient clearing. In the sharded state, model.parameters() exposes sharded parameters and their reduced .grad. In the unsharded state, it exposes unsharded parameters and their accumulation .grad. HSDP partial reduction buffers remain internal. Integer reshard_after_forward retains its existing temporary parameter layout without gradients. With CPU offload, unreduced gradients stay on the compute device and reduced shards are offloaded.
>
> Ordinary nn.Module.zero_grad() clears only the currently registered parameters. Clearing sharded gradients preserves hidden unsharded accumulation; clearing unsharded gradients preserves reduced history and internal HSDP partials. Gradient reduction consumes the unsharded .grad and sets it to None, while HSDP retains partial buffers until all-reduce completes. synchronize_gradients() completes pending reductions before global clipping or optimizer updates; callers retaining unsharded parameters also reshard before updating them. Conversion checks preserve native buffers on no-op conversions and explain how to complete reduction, reshard, and clear gradients before incompatible conversions. Pending gradients must be consumed or cleared on their actual owners before changing reduction policy.
>
> **Sharded gradient dtype:**
>
> These three rows fully determine `self.sharded_grad_dtype`. Neither mixed-precision field nor `self.unsharded_grad_dtype` changes this stored policy.
>
> | 1. Model parameter dtype before `fully_shard` | 2. Model parameter `grad_dtype` before `fully_shard` | 3. `self.sharded_grad_dtype` | 4. BC-breaking? Reason |
> | --- | --- | --- | --- |
> | `P` | Not set | `P` | **No.** Sharded dtype unchanged |
> | `P` | Explicit `G` | `G` | **Yes.** Sharded dtype follows `G` |
> | `P` | Explicit `None` | `None` | **Yes.** Fixed shard cast removed |
>
> **Unsharded gradient dtype and reduction:**
>
> Keep the original input `grad_dtype` in this table. `Not set` and explicit `G=P` both produce `self.sharded_grad_dtype=P`, but with `param_dtype=M` and no explicit `reduce_dtype`, their unsharded policies are `M` and `P`, respectively.
>
> For any actual parameter dtype `P`, all 12 combinations are listed below: three input gradient policies, two `param_dtype` choices, and two `reduce_dtype` choices. The BC column here covers only unsharded accumulation and reduction; sharded policy changes are covered above.
>
> | 1. Model parameter dtype before `fully_shard` | 2. Model parameter `grad_dtype` before `fully_shard` | 3. `mp_policy.param_dtype` | 4. `mp_policy.reduce_dtype` | 5. `self.unsharded_grad_dtype` | 6. BC-breaking? Reason |
> | --- | --- | --- | --- | --- | --- |
> | `P` | Not set | Explicit `None` / not set (resolves to `P`) | Explicit `None` / not set (resolves to `P`) | `P` | **No.** Accumulation and reduction unchanged |
> | `P` | Not set | Explicit `None` / not set (resolves to `P`) | Explicit `R` | `R` | **Yes.** Autograd accumulation uses `R` |
> | `P` | Not set | Explicit `M` | Explicit `None` / not set (resolves to `M`) | `M` | **No.** Accumulation and reduction unchanged |
> | `P` | Not set | Explicit `M` | Explicit `R` | `R` | **Yes.** Autograd accumulation uses `R` |
> | `P` | Explicit `G` | Explicit `None` / not set (resolves to `P`) | Explicit `None` / not set (resolves to `G`) | `G` | **Yes.** Accumulation and reduction use `G` |
> | `P` | Explicit `G` | Explicit `None` / not set (resolves to `P`) | Explicit `R` | `R` | **Yes.** Autograd accumulation uses `R` |
> | `P` | Explicit `G` | Explicit `M` | Explicit `None` / not set (resolves to `G`) | `G` | **Yes.** Accumulation and reduction use `G` |
> | `P` | Explicit `G` | Explicit `M` | Explicit `R` | `R` | **Yes.** Autograd accumulation uses `R` |
> | `P` | Explicit `None` | Explicit `None` / not set (resolves to `P`) | Explicit `None` / not set (resolves to actual gradient dtype) | `None` | **Yes.** Reduction follows actual gradient dtype |
> | `P` | Explicit `None` | Explicit `None` / not set (resolves to `P`) | Explicit `R` | `R` | **Yes.** Autograd accumulation uses `R` |
> | `P` | Explicit `None` | Explicit `M` | Explicit `None` / not set (resolves to actual gradient dtype) | `None` | **Yes.** Reduction follows actual gradient dtype |
> | `P` | Explicit `None` | Explicit `M` | Explicit `R` | `R` | **Yes.** Autograd accumulation uses `R` |
>
> Test Plan
>
> All 52 focused tests pass: 16 CPU/CUDA accumulation and unrestricted-policy cases, 23 mixed-precision cases, 12 HSDP pending-gradient cases, and one two-rank unused-parameter case. These exercise cancellation, per-parameter gradient dtypes, explicit reduction overrides, changing incoming gradient dtypes, partial-only synchronization, changing parameter subsets, and CPU offload. The new FP32 cancellation regression fails on the previous implementation. Lint and whitespace checks pass.
>
> ```bash
> CUDA_VISIBLE_DEVICES=0 /home/weif/.conda/envs/code-review/bin/python test/distributed/_composable/fsdp/test_fully_shard_conversion.py -k test_grad_dtype_default_reduction -k test_grad_dtype_none -v
> CUDA_VISIBLE_DEVICES=0,1,2,3 /home/weif/.conda/envs/code-review/bin/python test/distributed/_composable/fsdp/test_fully_shard_mixed_precision.py -k grad_dtype -v
> CUDA_VISIBLE_DEVICES=4,5,6,7 /home/weif/.conda/envs/code-review/bin/python test/distributed/_composable/fsdp/test_fully_shard_pending_grad.py -k test_grad_dtype_pending_reductions -k test_native_partial_ownership -v
> CUDA_VISIBLE_DEVICES=1,2 /home/weif/.conda/envs/code-review/bin/python test/distributed/_composable/fsdp/test_fully_shard_pending_grad.py -k test_grad_dtype_unused_parameters -v
> PATH=/home/weif/.conda/envs/code-review/bin:$PATH lintrunner -a
> git diff --check
> ```

Co-authored-by: Arkadip Maitra <amaitra@redhat.com>

### CODE REVIEW

File: torch/distributed/fsdp/_fully_shard/_fsdp_param.py
Comment (anshul-si) [inline torch/distributed/fsdp/_fully_shard/_fsdp_param.py:821]: I think this has to change now. Let's say we have param_dtype and reduce_dtype be bf16 and grad_dtype be fp32. reduce_dtype is set to None and then in https://github.com/pytorch/pytorch/blob/7b8d9471a4d929522190ff37b19fc3c1ebfea3e4/torch/distributed/fsdp/_fully_shard/_fsdp_collectives.py#L552-L560, reduce_dtype = reduce_dtype or grad_dtype, resulting in the wrong reduce_dtype

File: torch/distributed/fsdp/_fully_shard/_fsdp_collectives.py
Comment (anshul-si) [inline torch/distributed/fsdp/_fully_shard/_fsdp_collectives.py:758]: I don't think it affects computation, but new_sharded_dtensor_grad uses the original param sharding spec, which means param.grad._spec.tensor_meta.dtype = BF16. This wasn't a problem before since we always casted grad_dtype or orig_dtype.

File: torch/distributed/fsdp/_fully_shard/_fsdp_param.py [hunk 1]
```diff
@@ -1289,6 +1344,10 @@ def reset_sharded_param(self):
                     f"instead of {self.sharded_param}"
                 )
             self.sharded_param = new_param
+            # grad_dtype lives on the Parameter's autograd metadata and so
+            # survives an in-place `.data` swap, but a replacement Parameter
+            # object starts from the default and has to be re-stamped.
+            self._maybe_set_explicit_grad_dtype(self.sharded_param)
```
Comment (anshul-si) [inline torch/distributed/fsdp/_fully_shard/_fsdp_param.py:1350]: This might not be the right location for this. Module._apply() uses swap_tensors() for DTensor parameters, which preserves the Python object identity while replacing its TensorImpl/autograd metadata. Therefore, after model.to(...), new_param is self.sharded_param remains true and this re-stamping is skipped. The visible parameter defaults its grad_dtype to the converted parameter dtype, while
_explicit_grad_dtype retains the original setting; backward then fails when assigning the resulting gradient.

File: test/distributed/_composable/fsdp/test_fully_shard_mixed_precision.py [hunk 2]
```diff
@@ -458,6 +459,90 @@ def _reduce_1d_partial_grads(
...
+                        fsdp_param._unsharded_param.grad,
+                    ):
+                        if grad is not None:
+                            self.assertEqual(grad.dtype, expected_accumulate)
+
+        for param in model.parameters():
+            self.assertEqual(param.grad.dtype, grad_dtype)
+
+    @skip_if_lt_x_gpu(2)
+    def test_grad_dtype_preserved_across_load_state_dict(self):
```
Comment (anshul-si) [inline test/distributed/_composable/fsdp/test_fully_shard_mixed_precision.py:530]: I'm not sure this test is doing what it claims. load_state_dict will only install a replacement Parameter if assign = True. load_state_dict defaults assign = False, so it copies the state-dict value into the existing Parameter, so the parameter-replacement branch is not tested.

**Comment by anijain2305** [comment]
Agent Summary
----------------

# FSDP2 `grad_dtype` Follow-ups Needed for TorchTitan

I tested PyTorch commit `9b3e57d8fe1` (`[FSDP2] Honor uniform
Tensor.grad_dtype in fully_shard`) with TorchTitan's fused MXFP8 WGRAD
accumulation. The BF16-master / FP32-gradient / BF16-reduce configuration
works, but two details do not yet match the intended contract:

> `grad_dtype` owns gradient computation, local microbatch accumulation, and
> the dtype of sharded `param.grad`. `reduce_dtype` owns collective copy-in,
> reduce-scatter, and all-reduce only.

## 1. An explicitly configured dtype is lost when it equals the original dtype

`FSDPParam.__init__` currently uses:

```python
grad_dtype = param.grad_dtype
explicit_grad_dtype = (
    grad_dtype if param.requires_grad and grad_dtype != param.dtype else None
)
```

This cannot distinguish an explicit assignment from Tensor's default effective
dtype when the two values are equal.

That breaks TorchTitan's default configuration:

```text
training.dtype             = float32   # original/master parameter
training.grad_dtype        = float32   # explicitly requested gradient
training.mixed_precision_param = bfloat16
training.mixed_precision_reduce = float32
```

Titan explicitly sets `param.grad_dtype = torch.float32` before `fully_shard`,
but FSDP records `_explicit_grad_dtype = None`. The unsharded BF16 parameter
therefore gets BF16 `grad_dtype`. After the first no-sync microbatch, FSDP moves
its BF16 gradient into `unsharded_accumulated_grad` as FP32 and clears
`_unsharded_param.grad`. The next fused WGRAD cannot find the running
`weight.grad` into which it should accumulate.

The core implementation needs a way to preserve an explicitly assigned
`grad_dtype` even when it equals the original parameter dtype. Comparing the
effective `param.grad_dtype` with `param.dtype` is insufficient. A PyTorch test
should cover an FP32 original parameter with explicit FP32 `grad_dtype` and an
FSDP `param_dtype` of BF16.

## 2. Deferred accumulation still depends on `reduce_dtype`

**Human note: Ignore this one. We know that this is for BC**

`FSDPParam._accumulate_grad_dtype` currently selects the wider of
`grad_dtype` and `reduce_dtype`. Consequently, this configuration also moves
the gradient off the unsharded parameter after each no-sync backward:

```text
training.grad_dtype             = bfloat16
training.mixed_precision_reduce = float32
```

This means `reduce_dtype` still controls local microbatch accumulation. It also
prevents fused WGRAD accumulation because `weight.grad` is cleared between
microbatches.

For the agreed independent-knob contract, deferred unsharded accumulation must
remain in `grad_dtype`. The collective path should cast from `grad_dtype` to
`reduce_dtype` only when packing/copying into the reduce-scatter input. A test
should use `set_requires_gradient_sync(False)` for multiple backwards and
verify that `_unsharded_param.grad` remains present and in BF16 when
`grad_dtype=BF16` and `reduce_dtype=FP32`.

## Observed integration results

- BF16 original/master parameter, explicit FP32 `grad_dtype`, BF16
  `reduce_dtype`: passes; fused WGRAD remains in `weight.grad` and the second
  microbatch folds into it.
- FP32 original/master parameter, explicit FP32 `grad_dtype`, BF16 compute
  parameter, FP32 `reduce_dtype`: fails after the first microbatch because
  `weight.grad` is `None`.
- BF16 original/master parameter, BF16 `grad_dtype`, FP32 `reduce_dtype`: fails
  after the first microbatch because `weight.grad` is `None`.

The last two failures are independent: the first exercises loss of an explicit
same-as-original dtype, while the second exercises `reduce_dtype` influencing
local accumulation.

## Separate TorchTitan optimizer compatibility note

PyTorch fused Adam/AdamW currently rejects BF16 sharded parameters paired with
FP32 sharded gradients (`params, grads, exp_avgs, and exp_avg_sqs must have same
dtype, device, and layout`). This is separate from the FSDP issues above.
TorchTitan's existing full-BF16 recipes therefore need to set
`training.grad_dtype="bfloat16"` explicitly unless optimizer support for mixed
parameter/gradient dtypes is added.

File: torch/distributed/fsdp/_fully_shard/_fsdp_param.py [hunk 3]
```diff
@@ -199,7 +199,6 @@ class FSDPParam:
     _sharded_post_forward_param_data: torch.Tensor | None  # 1D
     _sharded_post_forward_param: nn.Parameter | None  # ND
     _unsharded_param: nn.Parameter  # ND
-    unsharded_accumulated_grad: torch.Tensor | None  # ND
```
Comment (weifengpy) [inline torch/distributed/fsdp/_fully_shard/_fsdp_param.py:202]: still available as `@property`

File: torch/distributed/fsdp/_fully_shard/_fsdp_collectives.py
Comment (weifengpy) [reply torch/distributed/fsdp/_fully_shard/_fsdp_collectives.py:788]: addresses this in to_sharded_dtensor by setting tensor_meta.dtype to the actual gradient dtype
